stripe-core/src/main/java/com/stripe/android/core/browser/customtabs/CustomTabsManager.kt

```diff
@@ -0,0 +1,221 @@
+            override fun onServiceConnected(
+                client: CustomTabsClient
+            ) {
+                log("Service connected")
+                this@CustomTabsManagerImpl.client = client
+                this@CustomTabsManagerImpl.client?.warmup(0)
+                session = this@CustomTabsManagerImpl.client?.newSession(CustomTabsCallback())
+            }
+
+            override fun onServiceDisconnected() {
+                log("Service disconnected")
+                client = null
+                session = null
+            }
+        })
+        bindCustomTabService(context)
+    }
+
+    /**
+     * Unbinds the Activity from the CustomTabsService.
+     */
+    override fun onStop(owner: LifecycleOwner) {
+        super.onStop(owner)
+        val context = owner as? Context ?: return
+        if (connection == null) return
+        connection?.let {
+            runCatching {
+                if (serviceBound) {
+                    context.unbindService(it)
+                    serviceBound = false
+                } else {
+                    log("OnStop: service was not bound")
+                }
+            }
+                .onFailure { log("OnStop: couldn't unbind, ${it.stackTraceToString()}") }
+                .onSuccess { log("OnStop: service unbound") }
+        }
+        client = null
+        connection = null
+        session = null
+    }
+
+    /**
+     * Creates or retrieves an exiting CustomTabsSession.
+     *
+     * @return a CustomTabsSession, or null if a connection to the service could not be established.
+     */
+    private fun getSession(): CustomTabsSession? {
+        when {
+            client == null -> session = null
+            session == null -> session = client?.newSession(null)
+        }
+        return session
+    }
+
+    override fun openCustomTab(
+        activity: Activity,
+        uri: Uri,
+        fallback: () -> Unit
+    ) {
+        runCatching {
+            val packageName: String? = getCustomTabsPackage(uri, activity)
+            val customTabsIntent = buildCustomTabsIntent()
+
+            // If we cant find a package name no browser that supports Custom Tabs is installed.
+            if (packageName == null) {
+                if (attemptToLaunchInNativeApp(activity, uri)) {
+                    log("uri launched on native app")
+                } else {
+                    log("Custom tabs unsupported, using fallback")
+                    fallback.invoke()
+                }
+            } else {
+                log("Opening Custom Tab with package: $packageName")
+                customTabsIntent.intent.setPackage(packageName)
+                customTabsIntent.launchUrl(activity, uri)
+            }
+        }.onFailure {
+            log("Failed to open Custom Tab, using fallback: ${it.stackTraceToString()}")
+            fallback.invoke()
+        }
+    }
+
+    @Suppress("SwallowedException")
```

Comment:
Should we still log the exception?

Reply:
https://github.com/stripe/stripe-android/pull/7081/commits/a297cef74f3e22fbda584fd4733622203a2e73db